context/opencensus-context/opencensus/common/runtime_context/__init__.py

```diff
@@ -0,0 +1,174 @@
+        """
+
+        raise NotImplementedError  # pragma: NO COVER
+
+    def apply(self, snapshot):
+        """Set the current context from a given snapshot dictionary"""
+
+        for name in snapshot:
+            setattr(self, name, snapshot[name])
+
+    def snapshot(self):
+        """Return a dictionary of current slots by reference."""
+
+        return dict((n, self._slots[n].get()) for n in self._slots.keys())
+
+    def __repr__(self):
+        return ('{}({})'.format(type(self).__name__, self.snapshot()))
+
+    def __getattr__(self, name):
+        if name not in self._slots:
+            raise AttributeError('{} is not a registered context slot'
+                                 .format(name))
+        slot = self._slots[name]
+        return slot.get()
+
+    def __setattr__(self, name, value):
+        if name not in self._slots:
+            raise AttributeError('{} is not a registered context slot'
+                                 .format(name))
+        slot = self._slots[name]
+        slot.set(value)
+
+    def with_current_context(self, func):
+        """Capture the current context and apply it to the provided func"""
+
+        caller_context = self.snapshot()
+
+        def call_with_current_context(*args, **kwargs):
+            try:
+                backup_context = self.snapshot()
+                self.apply(caller_context)
+                return func(*args, **kwargs)
+            finally:
+                self.apply(backup_context)
+
+        return call_with_current_context
+
+
+class _ThreadLocalRuntimeContext(_RuntimeContext):
+    _lock = threading.Lock()
+    _slots = {}
+
+    class Slot(object):
+        _thread_local = threading.local()
+
+        def __init__(self, name, default):
+            self.name = name
+            self.default = default if callable(default) else (lambda: default)
+
+        def clear(self):
+            setattr(self._thread_local, self.name, self.default())
+
+        def get(self):
+            try:
+                return getattr(self._thread_local, self.name)
+            except AttributeError:
+                value = self.default()
+                self.set(value)
+                return value
+
+        def set(self, value):
+            setattr(self._thread_local, self.name, value)
+
+    @classmethod
+    def clear(cls):
+        with cls._lock:
+            for name in cls._slots:
+                slot = cls._slots[name]
+                slot.clear()
+
+    @classmethod
+    def register_slot(cls, name, default=None):
+        with cls._lock:
+            if name in cls._slots:
+                raise ValueError('slot {} already registered'.format(name))
+            slot = cls.Slot(name, default)
+            cls._slots[name] = slot
+            return slot
+
+
+class _AsyncRuntimeContext(_RuntimeContext):
+    _lock = threading.Lock()
+    _slots = {}
+
+    class Slot(object):
+        def __init__(self, name, default):
+            import contextvars
```

Comment:
Why bury the import in here?

Reply:
I was trying to reduce the visibility of contextvars to avoid accidental use.
Alternatively we could use if/else and import at the top level (which seems harder to maintain?)